### PR TITLE
Add core dump generation

### PR DESCRIPTION
**Pull request by eerhardt**
We should be able to run the OSX leg of the PR over and over until we see the issue.

@gkhanna79 @jkotas @stephentoub @JohnChen0 @sergiy-k @janvorli

### CODE REVIEW

File: build_projects/dotnet-cli-build/build.sh [hunk 1]
```diff
@@ -101,6 +101,15 @@ then
     ulimit -n 1024
 fi
 
+# Temporary logic to turn core dumps on so we can catch exit code 138 errors on OSX
+if [ "$(uname -s)" == "Darwin" ]; then
+  # If there are no core files already in /cores/, we'll enable core dump
+  # generation for this shell.
+  if [ "$(ls -A /cores)" ]; then
```
Comment (gkhanna79) [inline build_projects/dotnet-cli-build/build.sh]: @sergiy-k Why would we need this check and not enable unconditionally?

Comment (gkhanna79) [inline build_projects/dotnet-cli-build/build.sh]: @eerhardt Please do the reverse (i.e. disable the generation) once the build script ends and do a "ls -l /cores" as well.

Comment (eerhardt) [inline build_projects/dotnet-cli-build/build.sh]: > Please do the reverse (i.e. disable the generation) once the build script ends

Is this necessary?  @adityamandaleeka didn't do the reverse in the buildtools change.

> and do a "ls -l /cores" as well.

Can you explain a little more what you want here?  What do you want to see happening?

Comment (gkhanna79) [inline build_projects/dotnet-cli-build/build.sh]: > Is this necessary?

We do not want crash dumps to be generated once the CLI build exits.

> What do you want to see happening?

Capture contents of cores folder (before and after) in the log to determine if we caught anything or not.

**Comment by janvorli** [comment]
I was able to repro it locally on my OSX today. I have left it running in a loop for few hours and then there were two core files. Both had the same call stack, but since the release binaries have stripped symbols and I wasn't able to make lldb load the .dwarf symbol files, I was missing locals etc. So I have rebuilt the coreclr with the stripping and optimizations disabled (by patching the cmake / compiler options files) and copied that over the files in the dotnet tool subfolder. I am waiting for another repro with that.
This are top 10 frames from the stack:
```
(lldb) bt
* thread #1: tid = 0x0000, 0x000000010778a329 libcoreclr.dylib`AssemblyBindingHolder::CreateAssemblyBinding(LoaderHeap*) + 169, stop reason = signal SIGSTOP
  * frame #0: 0x000000010778a329 libcoreclr.dylib`AssemblyBindingHolder::CreateAssemblyBinding(LoaderHeap*) + 169
    frame #1: 0x00000001077891c3 libcoreclr.dylib`AssemblySpecBindingCache::StoreFile(AssemblySpec*, PEAssembly*) + 259
    frame #2: 0x00000001075d128e libcoreclr.dylib`AppDomain::AddFileToCache(AssemblySpec*, PEAssembly*, int) + 62
    frame #3: 0x00000001075d2721 libcoreclr.dylib`AppDomain::BindAssemblySpec(AssemblySpec*, int, int, StackCrawlMark*, AssemblyLoadSecurity*, int) + 1185
    frame #4: 0x0000000107713c20 libcoreclr.dylib`PEFile::LoadAssembly(unsigned int, IMDInternalImport*, char const*, char const*) + 352
    frame #5: 0x00000001075f50e8 libcoreclr.dylib`Module::LoadAssembly(AppDomain*, unsigned int, char const*, char const*) + 344
    frame #6: 0x00000001075e4add libcoreclr.dylib`Assembly::FindModuleByTypeRef(Module*, unsigned int, Loader::LoadFlag, int*) + 397
    frame #7: 0x000000010760902c libcoreclr.dylib`ClassLoader::LoadTypeDefOrRefThrowing(Module*, unsigned int, ClassLoader::NotFoundAction, ClassLoader::PermitUninstantiatedFlag, unsigned int, ClassLoadLevel) + 444
    frame #8: 0x000000010772b395 libcoreclr.dylib`SigPointer::GetTypeHandleThrowing(Module*, SigTypeContext const*, ClassLoader::LoadTypesFlag, ClassLoadLevel, int, Substitution const*, ZapSig::Context const*) const + 1797
    frame #9: 0x000000010777132f libcoreclr.dylib`ZapSig::DecodeType(Module*, Module*, unsigned char const*, ClassLoadLevel) + 111
    frame #10: 0x000000010771f38e libcoreclr.dylib`DynamicHelperFixup(TransitionBlock*, unsigned long*, unsigned int, Module*, CORCOMPILE_FIXUP_BLOB_KIND*, TypeHandle*, MethodDesc**, FieldDesc**) + 926
```
The failure point is here:
```asm
->  0x10778a329 <+169>: movq   $0x0, 0x70(%rcx)
```
The rcx + 0x70 seems to be a valid 16 byte aligned address in the dump, so the only thought that I have is that somehow the memory protection was set to read only (or maybe even PROT_NONE if the debugger would still show it from the dump in that case)

**Comment by gkhanna79** [comment]
This is agreat @janvorli. Yesterday, the issue you ran into with checked CoreCLR was https://github.com/dotnet/coreclr/pull/5732, which is now fixed.

CC @JohnChen0 @fadimounir @jkotas since this is happening via DynamicHelperFixup.

**Comment by eerhardt** [comment]
@janvorli - The CLI always uses the release build of CoreCLR. So if we get a core dump on a Jenkins machine, will it so us any good?  Will we be able to debug it?

**Comment by janvorli** [comment]
@gkhanna79 great that it was fixed, I will then try to repro it with debug build too.

**Comment by jkotas** [comment]
https://github.com/dotnet/coreclr/issues/5721 is around the same `AssemblySpecBindingCache` that where this AV is getting hit. I am wondering whether they have something in common.

**Comment by eerhardt** [comment]
test OSX x64 Release Build

File: build_projects/dotnet-cli-build/build.sh (hunk 1 above)
Comment (adityamandaleeka) [inline build_projects/dotnet-cli-build/build.sh]: @eerhardt @gkhanna79 

The reason why the BuildTools change doesn't reset ulimit is that each of those scripts is invoked in its own shell and the ulimit change is limited to the shell it's made in (and any children). I haven't looked into how this CLI script is invoked or by whom, but that may or may not apply here.

**Comment by eerhardt** [comment]
test OSX x64 Release Build

This failed twice in a row on my other PR - and now it won't fail once...

**Comment by eerhardt** [comment]
test RHEL7.2 x64 Release Build
test OSX x64 Release Build

**Comment by janvorli** [comment]
Strange - I've synced to the latest coreclr and I wasn't able to repro the issue anymore - running in a loop for at least two hours.

**Comment by eerhardt** [comment]
test OSX x64 Release Build

**Comment by eerhardt** [comment]
@janvorli - that's good, right?  ;)

I know we just took a fix for https://github.com/dotnet/coreclr/pull/5732 in release/1.0.0.  Which branch are you in?

**Comment by eerhardt** [comment]
test OSX x64 Release Build

**Comment by eerhardt** [comment]
We got what we needed out of this.

**Comment by eerhardt** [comment]
@gkhanna79 @jkotas @stephentoub @JohnChen0 @sergiy-k @janvorli @adityamandaleeka  - I've re-opened this PR with the new logic for Linux.  I have made all of you collaborators on my fork, so you can push any change you want to this branch, if the logic isn't correct.  I may not be around very much today.

One place I didn't know exactly to put on this PR was passing in the executable name into `print_info_from_core_file`. The `dotnet` executable that is used at test time is contained under `$RepoRoot/artifacts/$RID/stage2`, I didn't know if I needed to pass the whole path to the executable or not.  There is a version of `dotnet` on the path which should match, but it isn't necessarily the path to the executable that failed.

**Comment by janvorli** [comment]
@eerhardt - when I am trying to repro it locally, which cli branch should I use? I was using rel/1.0.0-preview2.

**Comment by eerhardt** [comment]
Yes that is the correct branch where preview2 will be released from. It's the same branch that this PR is against.

**Comment by eerhardt** [comment]
@dotnet-bot test this please

**Comment by eerhardt** [comment]
test CentOS7.1 x64 Debug Build
test RHEL7.2 x64 Release Build
test Debian8.2 x64 Debug Build
test OpenSUSE13.2 x64 Debug Build
test Ubuntu x64 Release Build
test Ubuntu16.04 x64 Debug Build

**Comment by eerhardt** [comment]
Test Fedora23 x64 Debug Build

**Comment by gkhanna79** [comment]
LGTM

**Comment by eerhardt** [comment]
Looks like the RHEL leg failed with an exit code 139, but the core dump logging didn't log anything that I can see. Can someone take a look?

**Comment by stephentoub** [comment]
I see:
```
Exit Code: 139
      StdOut:
      
      StdErr:
      
      
      Stack Trace:
           at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message)
           at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
        /mnt/resource/j/workspace/dotnet_cli/rel_1.0.0-preview2/release_rhel7.2_x64_prtest/test/Microsoft.DotNet.Tools.Tests.Utilities/Assertions/CommandResultAssertions.cs(32,0): at Microsoft.DotNet.Tools.Test.Utilities.CommandResultAssertions.Pass()
        /mnt/resource/j/workspace/dotnet_cli/rel_1.0.0-preview2/release_rhel7.2_x64_prtest/test/dotnet-publish.Tests/PublishStandaloneTests.cs(39,0): at Microsoft.DotNet.Tools.Publish.Tests.PublishStandaloneTests.StandaloneAppHasResourceDependency()
    Microsoft.DotNet.Tools.Publish.Tests.PublishStandaloneTests.StandaloneAppHasResourceDependency [FAIL]
```
Does that means it failed due to an unhandled exception?

**Comment by eerhardt** [comment]
That's the random exit code we've been trying to capture. I had the core dump turned on, but it didn't look like the logging code logged anything.

**Comment by stephentoub** [comment]
> I had the core dump turned on, but it didn't look like the logging code logged anything.

I'm wondering about the stack trace that was captured... where did that come from?

**Comment by gkhanna79** [comment]
Yes, that usually comes when managed exception goes unhandled.

**Comment by stephentoub** [comment]
> that usually comes when managed exception goes unhandled.

What exit code do we normally use when we crash due to an unhandled exception?  Is it that we normally use a different one, and something is causing us to seg fault while we're in the process of crashing due to the unhandled exception?

**Comment by janvorli** [comment]
Unhandled exception aborts, so the exit code should be 134.

**Comment by sergiy-k** [comment]
test OSX x64 Release Build
test OSX x64 Debug Build
test CentOS7.1 x64 Debug Build
test CentOS7.1 x64 Release Build
test RHEL7.2 x64 Debug Build

**Comment by sergiy-k** [comment]
@dotnet-bot test RHEL7.2 x64 Release Build

**Comment by eerhardt** [comment]
We got what we needed out of this (for real this time).

**Pull request closed without merging**